src/main/java/org/fungover/mmotodo/service/AuthService.java

```diff
@@ -0,0 +1,151 @@
+
+    }
+
+
+    public boolean performLogout() {
+        String accessTokenValue = retrieveAccessToken();
+
+        if (accessTokenValue != null) {
+            String basicAuth = Base64.getEncoder().encodeToString((clientId + ":" + clientSecret).getBytes());
+            HttpHeaders headers = new HttpHeaders();
+            headers.set("Authorization", "Basic " + basicAuth);
+            headers.setContentType(MediaType.APPLICATION_JSON);
+
+            String jsonBody = "{\"access_token\":\"" + accessTokenValue + "\"}";
+            HttpEntity<String> request = new HttpEntity<>(jsonBody, headers);
+
+            // Send request to sign out from OAuth application
+            try {
+                new RestTemplate().exchange(
+                        githubDeleteUrl,
+                        HttpMethod.DELETE,
+                        request,
+                        String.class
+                );
+
+                clearAuthenticationCookies();
+                return true;
+            } catch (HttpClientErrorException.UnprocessableEntity e) {
+                logger.error("Couldn't sign out from GitHub: ", e.getMessage());
+            } catch (Exception e) {
+                logger.error("Exception during logout: ", e.getMessage());
+            }
+        } else {
+            // Handle the case where access token retrieval fails
+            logger.warn("Couldn't find access token");
+        }
+
+        return false;
+    }
+
+
+    public String retrieveAccessToken() {
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+
+        if (auth instanceof OAuth2AuthenticationToken) {
+
+            // Getting the access token to make a request to sign out from account
+            OAuth2AuthenticationToken token = (OAuth2AuthenticationToken) auth;
+            String clientRegistrationId = token.getAuthorizedClientRegistrationId();
+            OAuth2AuthorizedClient authorizedClient = authorizedClientService.loadAuthorizedClient(
+                    clientRegistrationId, token.getName());
+
+            OAuth2AccessToken accessToken = authorizedClient.getAccessToken();
+
+
+            return accessToken.getTokenValue();
+        }
+        return null;
+    }
+
+
+    private void clearAuthenticationCookies() {
+        ServletRequestAttributes attributes = (ServletRequestAttributes) RequestContextHolder.getRequestAttributes();
+
+        if (attributes == null) {
+            return;
+        }
+
+        HttpServletRequest req = attributes.getRequest();
+        HttpServletResponse res = attributes.getResponse();
+
+        Cookie[] cookies = req.getCookies();
+
+        for (Cookie cookie : cookies) {
+            if (cookie.getName().equals("JSESSIONID")) {
+                cookie.setMaxAge(0);
+                cookie.setPath("/");
+                assert res != null;
+                res.addCookie(cookie);
+            }
+        }
+    }
```

Comment:
In this snippet, you have the null check late with `assert`. I think a better way to do it (to ensure no null pointer exceptions) would be to move the null check before you add the cookies to an array. Something similar to this:

```
if (req != null && res != null) {
            Cookie[] cookies = req.getCookies();

            if (cookies != null) {
                for (Cookie cookie : cookies) {
                   // if cookie name equals ...
                    }
                }
            }
```